app/src/org/commcare/android/adapters/HomeScreenAdapter.java

```diff
@@ -1,150 +1,209 @@
+
+        if (viewType == TYPE_BUTTON) {
+            View layoutView = inflater.inflate(R.layout.home_card, parent, false);
+            return new SquareButtonViewHolder(layoutView);
+        } else if (viewType == TYPE_HEADER) {
+            View header = inflater.inflate(R.layout.grid_header_top_banner, parent, false);
+            return new HeaderViewHolder(header);
         }
+
+        throw new RuntimeException("no " + viewType +
+                " type exists, should be TYPE_BUTTON or TYPE_HEADER");
     }
 
-    /**
-     * Sets the onClickListener for the given button
-     *
-     * @param resourceCode Android resource code (R.id.$button or R.layout.$button)
-     * @param listener     OnClickListener for the button
-     */
-    public void setOnClickListenerForButton(int resourceCode, View.OnClickListener listener) {
-        int buttonIndex = getButtonIndex(resourceCode);
-        SquareButtonWithNotification button = (SquareButtonWithNotification) getItem(buttonIndex);
-        if (button != null) {
-            if (BuildConfig.DEBUG) {
-                Log.d(TAG, "Preexisting button when calling setOnClickListenerForButton");
-            }
-            button.setOnClickListener(listener);
-        } else {
-            if (BuildConfig.DEBUG) {
-                Log.d(TAG, "Button did not exist when calling setOnClickListenerForButton!");
-            }
+    @Override
+    public void onBindViewHolder(RecyclerView.ViewHolder holder, int i) {
+        if (holder instanceof SquareButtonViewHolder) {
+            bindCard((SquareButtonViewHolder)holder, i, null);
+        } else if (holder instanceof HeaderViewHolder) {
+            bindHeader((HeaderViewHolder)holder);
         }
     }
 
-    public SquareButtonWithNotification getButton(int resourceCode) {
-        return buttons[getButtonIndex(resourceCode)];
+    @Override
+    public void onBindViewHolder(RecyclerView.ViewHolder holder,
+                                 int i, List<Object> payload) {
+        if (holder instanceof SquareButtonViewHolder) {
+            bindCard((SquareButtonViewHolder)holder, i, payload);
+        } else if (holder instanceof HeaderViewHolder) {
+            bindHeader((HeaderViewHolder)holder);
+        }
     }
 
-    public void setNotificationTextForButton(int resourceCode, String notificationText) {
-        SquareButtonWithNotification button = getButton(resourceCode);
-        if (button != null) {
-            button.setNotificationText(notificationText);
-            notifyDataSetChanged();
+    private void bindCard(SquareButtonViewHolder squareButtonViewHolder,
+                          int i, List<Object> payload) {
+        HomeCardDisplayData cardDisplayData = getItem(i);
+        String notificationText = null;
+
+        if (payload != null) {
+            notificationText = getLastPayloadString(payload);
         }
+
+        cardDisplayData.textSetter.update(cardDisplayData,
```

Comment:
Very minor, but any reason the code for setting the view holder's text isn't just part of the setupViewHolder() method? Seems like it would logically be a part of that.